waku/v2/api/filter.go

```diff
@@ -67,23 +66,26 @@ func (apiSub *Sub) Unsubscribe() {
 
 }
 
-func (apiSub *Sub) healthCheckLoop() {
-	// Health checks
-	ticker := time.NewTicker(FilterPingTimeout)
-	defer ticker.Stop()
+func (apiSub *Sub) waitOnSubClose() {
 	for {
 		select {
 		case <-apiSub.ctx.Done():
 			apiSub.log.Debug("healthCheckLoop: Done()")
 			apiSub.cleanup()
 			return
-		case <-ticker.C:
-			apiSub.log.Debug("healthCheckLoop: checkAliveness()")
-			topicCounts := apiSub.getTopicCounts()
-			apiSub.resubscribe(topicCounts)
+		case subId := <-apiSub.closing:
+			//trigger closing and resubscribe flow for subscription.
+			apiSub.closeAndResubscribe(subId)
 		}
 	}
+}
+
+func (apiSub *Sub) closeAndResubscribe(subId string) {
+	apiSub.log.Debug("sub closeAndResubscribe", zap.String("subID", subId))
 
+	apiSub.subs[subId].Close()
+	apiSub.resubscribe()
+	delete(apiSub.subs, subId)
```

Comment:
Since access to this is protected by context, and only operations on subs is via the single go-routine and the multiplex go-routine gets invoked only once these operations are completed. 

I have also moved delete above resubscribe to ensure no race occurs. 
Do you still see a challenge that can cause race?

Reply:
I guess it should be okay. Worst case scenario, we notice this in one of the test-with-race runs :)